### DISCUSSION TITLE
Rust error reporting does not work correctly

### DISCUSSION THREAD

**Discussion by vmsh0** [w0rp/ale#4120] [Q&A - Ask for help with problems]
Hi,

I'm using the plugin with the default settings in the latest neovim. It works fine for most languages, but I'm seeing some weird behaviour with Rust.

In short, it looks like ALE is taking error information from two different sources, as sometimes programs report the same errors as I would get when compiling with `cargo build`, and sometimes they report something entirely different. That is usually fixed by doing Ctrl-Z, giving `cargo build` manually in the project directory, going back to neovim, and forcing a refresh of ALE with `:w`.

As a practical example, I'm not writing a small client program for a REST API. I'm working at a proof-of-concept, which consists of a single file. At the top of the file I have
```
  use google_drive3::{DriveHub, oauth2, hyper, hyper_rustls};
>>use simple_error::bail;

  use std::error::Error;
  use std::fs::File;
  use std::io::BufReader;
```

As you can see, ALE reports an error at line 2 (and corresponding errors wherever the `bail` macro is used in the file). However, the program compiles just fine - and right after compiling and refreshing, as mentioned, ALE correctly reports no errors. But then if I change as little as some whitespace somewhere in the file, ALE goes back to reporting the "false" errors.

I don't know how ALE works internally, but I can imagine that if there's a fresh compilation result it gets its information for that - thus getting it right after a `cargo build`. But when the source file is newer than the last compile, it does something different - and apparently gets it wrong.

Can someone help me figure this out? I really don't know where to look.

**Comment by hsanson** [comment]
Start by inspecting the output of ALEInfo. It may provide some insights on what is going on.

**Reply by vmsh0** [reply]
Thank you. I did that, but either it doesn't say anything useful or I don't know what I need to be looking for.

For what is worth, even when the false errors are reported, `ALEInfo` reports output from `cargo check` that does _not_ contain the errors. I.e. ALE is getting them from somewhere else.

Do you need a copy of the output of `ALEInfo`? How can I obtain that?

**Reply by hsanson** [reply]
The ALEInfo output may help, use :ALEInfoToClipboard to copy it.

I do not use rust or cargo so the help I can provide is limited. Since you mention errors seem to be comming from somewhere else I recommend you create a minimal vimrc file with only ALE and test using it.

**Reply by vmsh0** [reply]
For some reason the output given by :ALEInfoToClipboard included more information. In particular, it mentioned the use of `rls`, which was after all the issue. I updated `rls` and now everything works as expected.

Thank you for you help!

**Reply by vmsh0** [reply]
As it turns out that was just a coincidence. The issue is presenting again. While the issue is present, I use `ALEInfoToClipboard`, and grep the resulting output for the errors ALE is showing for my code. It turns out those errors are not shown in ALEInfo. I have no idea what is up with that.

You can see my ALEInfo here: https://paste.sr.ht/blob/2d1a8e51f5981570360f08ed166e5349d236dd19

My `init.vim` looks like this:
```call plug#begin('~/.vim/plugged')

Plug 'lervag/vimtex'

let g:ale_completion_enabled = 1
Plug 'dense-analysis/ale'

call plug#end()

nnoremap <C-k> :ALEHover<CR>

au BufRead,BufNewFile *.asm set filetype=nasm```

I don't think there's anything in here that might be interfering with ALE.

**Comment by rymdbar** [comment]
This ticket appears to be specific to rls, which was deprecated in #4461. The general advice to anyone still encountering this issue is to switch to rust-analyzer, as the problem is unlikely to ever be resolved.

Given the above, could this discussion perhaps please be renamed to something along the lines of:

> Error reporting with rls does not work correctly

?
